### PR TITLE
Refactor XDR serialization/deserialization logic

### PR DESCRIPTION
**Pull request by orbitlens**
- Refactor XDR serialization/deserialization logic.
- Replace `long` dependency with native `BigInt` arithmetics.
- Replace `lodash` dependency with built-in Array and Object methods, iterators.
- Add `buffer` dependency for WebPack browser polyfill.
- Update devDependencies to more recent versions, modernize bundler pipeline.
- Automatically grow underlying buffer on writes (#84 fixed).
- Always check that the entire read buffer is consumed (#32 fixed).
- Check actual byte size of the string on write (#33 fixed).
- Fix babel-polyfill build warnings (#34 fixed).
- Bump major version.

----------

Minified browser bundle size decreased from **335 KB** (prev version) to **43 KB** (this version). Memory consumption during serialization decreased due to the smaller writer buffer allocation.

We already actively use it as a drop-in replacement for original `js-xdr` package for ingestion. So far, no compatibility issues have been found.
 
Also I created a primitive performance benchmark for Node.js to test serialization/deserialization speed for refactored version vs original. It shows significant performance boost, up to 5X depending on the serialized type.

[Benchmark link](https://github.com/orbitlens/js-xdr-perf-test). Benchmark results on my computer: 

```
Performance benchmark
10 samples 10000 iterations each
------------------------------------------------
LedgerHeader.fromXDR - new: 0.0435s (min 0.038s - max 0.068s)
LedgerHeader.fromXDR - old: 0.272s (min 0.264s - max 0.302s)
difference: 6.3X

LedgerHeader.toXDR - new: 0.0698s (min 0.06s - max 0.101s)
LedgerHeader.toXDR - old: 0.348s (min 0.263s - max 0.558s)
difference: 5.0X

TransactionEnvelope.fromXDR - new: 0.423s (min 0.412s - max 0.477s)
TransactionEnvelope.fromXDR - old: 1.77s (min 1.701s - max 1.918s)
difference: 4.2X

TransactionEnvelope.toXDR - new: 0.436s (min 0.342s - max 0.715s)
TransactionEnvelope.toXDR - old: 0.646s (min 0.613s - max 0.695s)
difference: 1.5X

TransactionMeta.fromXDR - new: 3.20s (min 2.694s - max 4.016s)
TransactionMeta.fromXDR - old: 9.76s (min 9.543s - max 10.194s)
difference: 3.1X

TransactionMeta.toXDR - new: 2.35s (min 2.149s - max 2.549s)
TransactionMeta.toXDR - old: 3.19s (min 2.934s - max 3.46s)
difference: 1.4X

TransactionResult.fromXDR - new: 0.0370s (min 0.034s - max 0.056s)
TransactionResult.fromXDR - old: 0.0627s (min 0.059s - max 0.074s)
difference: 1.7X

TransactionResult.toXDR - new: 0.0626s (min 0.057s - max 0.07s)
TransactionResult.toXDR - old: 0.218s (min 0.184s - max 0.272s)
difference: 3.5X
```

### CODE REVIEW

File: src/struct.js [hunk 1]
```diff
@@ -1,72 +1,76 @@
...
     ChildStruct.structName = name;
 
     context.results[name] = ChildStruct;
 
-    ChildStruct._fields = fields.map(([fieldName, field]) => {
+    const mappedFields = new Array(fields.length);
+    for (let i = 0; i < fields.length; i++) {
+      const fieldDescriptor = fields[i];
+      const fieldName = fieldDescriptor[0];
+      let field = fieldDescriptor[1];
```
Comment (Shaptic) [inline src/struct.js:54]: Would deconstruction work here?

```suggestion
      let [ fieldName, field ] = fields[i];
```

**Review by Shaptic — APPROVED** [review]
Wow, Herculean stuff :clap: It'd be nice if we had .x -> .js tests to see if there were any output changes, but since you're using this for ingestion that gives really high confidence!

**Comment by orbitlens** [comment]
> Wow, Herculean stuff 👏 It'd be nice if we had .x -> .js tests to see if there were any output changes.

I tried generating `.x -> .js` with `examples/test.x` (via `yarn run test-generate`), but of course it makes sense to test with **.x** files from stellar-core. Could you please check the generated output for the difference? Because I'm a bit lost between various versions of `stellar-core` and curr/next XDR.

File: src/struct.js (hunk 1 above)
Comment (orbitlens) [reply src/struct.js:54]: That's exactly how I coded it from the first attempt. However, the linter was against it – in the linter's opinion, the `fieldName` is never changed, so it should be defined as `const`. I'm 💯 in favor of the proposed change, but the linter should be suppressed for the line in this case.

**Comment by tomerweller** [comment]
Great work. thanks @orbitlens!

**Pull request merged**